### DISCUSSION TITLE
CAN bus python library module for ESP32 (particularly ESP32-C3)

### DISCUSSION THREAD

**Discussion by RufusVS** [micropython/micropython#10535] [ESP32]
The ESP32 processor has native CAN capability but there is no support for it in micropython. I noticed some other chips supported by MicroPython DO have a CAN class library, and there must be some C code to drive the CAN on the chip as well.  Is anyone aware of work to this end?  In particular, I just picked up some QTPY ESP32-C3 board from Adafruit, and I'd sure like to make a CAN controller out of it.  (I'm aware I'd need a transceiver chip for true CAN but that's a secondary consideration).  I'd rather not use Arduino and C to program it.
(Although if anyone has links to C code for ESP32 CAN, please share and I might look into porting it myself, personal bandwidth permitting).
I'm aware of I2C can interfaces but I was hoping to use native CAN of the ESP32 chip.

**Comment by 2dof** [comment]
see: https://github.com/orgs/micropython/discussions/9584
 https://github.com/micropython/micropython/pull/9532

also see Awesome CAN Bus: https://github.com/iDoka/awesome-canbus  
in the "libraries" section You will find Libraries  in c ( [iso15765-canbus](https://github.com/devcoons/iso15765-canbus) - Implementation of ISO15765-2 in C. )

**Comment by mattytrentini** [comment]
Coincidentally, I've been looking in to CAN (and, more specifically, [CANopen](https://en.wikipedia.org/wiki/CANopen)) on MicroPython recently for a project at work.

The short answer is no, it's currently only implemented for STM32 in [pyb.CAN](https://docs.micropython.org/en/latest/library/pyb.CAN.html). That API is quite rich but modelled around the STM32 features so may be challenging to implement on other ports.

It would be a _very_ welcome feature to define a CAN API generically in `machine` and implement it for as many ports as possible - STM32 and ESP32 would be a great start.

CircuitPython has [canio](https://docs.circuitpython.org/en/latest/shared-bindings/canio/index.html); modelling the API on that would be an excellent start!

**Reply by jimmo** [reply]
See also https://github.com/micropython/micropython/pull/9532
